src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,342 @@
+    r = sol_flow_send_string_packet(node,
+        SOL_FLOW_NODE_TYPE_NETCTL_LIST_SERVICES__OUT__OUT,
+        name);
+    if (r < 0)
+        SOL_WRN("Failed to send network service name: %s", name);
+
+    sol_ptr_vector_append(&mdata->service_list, service);
+}
+
+void
+error_cb(void *data, const struct sol_netctl_service *service,
+    unsigned int error)
+{
+    const char *name;
+
+    name = sol_netctl_service_get_name(service);
+
+    if (name)
+        SOL_ERR("Service %s error is %d", name, error);
+}
+
+static int
+open_list_services(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    static const char *errmsg = "Could not add service monitor";
+    struct list_service_data *mdata = data;
+
+    mdata->node = node;
+
+    sol_ptr_vector_init(&mdata->service_list);
+
+    if (sol_netctl_add_service_monitor(network_service_cb, mdata))
+        goto error;
+
+    return 0;
+
+error:
+    sol_flow_send_error_packet(node, EINVAL, "%s", errmsg);
+    SOL_WRN("%s", errmsg);
+
+    return -EINVAL;
+}
+
+static void
+close_list_services(struct sol_flow_node *node, void *data)
+{
+    static const char *errmsg = "Could not delete service monitor";
+    struct list_service_data *mdata = data;
+
+    if (sol_netctl_del_service_monitor(network_service_cb, mdata))
+        goto error;
+
+    sol_ptr_vector_clear(&mdata->service_list);
+
+    return;
+
+error:
+    sol_flow_send_error_packet(node, EINVAL, "%s", errmsg);
+    SOL_WRN("%s", errmsg);
+}
+
+static int
+connection_handle(struct sol_flow_node *node, void *data, bool connect)
+{
+    struct connection_service_data *mdata = data;
+    const struct sol_ptr_vector *service_list;
+    struct sol_netctl_service *service;
+    uint16_t i = 0;
+    enum sol_netctl_service_state state;
+    int r;
+    const char *name;
+
+    static const char *state_msgs[] = {
+        [SOL_NETCTL_SERVICE_STATE_UNKNOWN] = "Unknown",
+        [SOL_NETCTL_SERVICE_STATE_IDLE] = "Idle",
+        [SOL_NETCTL_SERVICE_STATE_ASSOCIATION] = "Association",
+        [SOL_NETCTL_SERVICE_STATE_CONFIGURATION] = "Configuration",
+        [SOL_NETCTL_SERVICE_STATE_READY] = "Ready",
+        [SOL_NETCTL_SERVICE_STATE_ONLINE] = "Online",
+        [SOL_NETCTL_SERVICE_STATE_DISCONNECT] = "Disconnect",
+        [SOL_NETCTL_SERVICE_STATE_FAILURE] = "Failure",
+        [SOL_NETCTL_SERVICE_STATE_REMOVE] = "Remove",
+    };
+
+    service_list = sol_netctl_get_services();
+
+    while (service = sol_ptr_vector_get(service_list,i)) {
+        state = sol_netctl_service_get_state(service);
+        name = sol_netctl_service_get_name(service);
+
+        if (name && strcmp(name, mdata->service_name) == 0) {
+            if (state == SOL_NETCTL_SERVICE_STATE_IDLE && connect) {
+                sol_netctl_service_connect(service);
+            } else if ((state == SOL_NETCTL_SERVICE_STATE_READY) && (!connect)) {
+                  sol_netctl_service_disconnect(service);
+            }
```

Comment:
a note for the logic above: I don't think you should need to check the current state before doing the operating. Let the backend handle that. If you ask connect and it's already connected, or disconnect and it's already disconnected, the backend should handle that.

Eventually the backend will return an error "already connected" or "already disconnected", but that's harmless.

The logic you do here is prone to race condition. Suppose you look for IDLE && connect, but the service was in `CONNECTED` and few milliseconds later would go back to `IDLE` because a disconnect request was in progress. You'd miss the connect call. This would happen with a quick disconnect/connect cycle.

IOW: just call `sol_netctl_service_connect()` or `sol_netctl_service_disconnect()`. Then this logic can be moved in the *caller* of this function, instead of using a parameter. This function would be converted into a simple `find_service_by_name()` that does the loop and return the `service` handle. (more on that below, you can't get the state right away)